bob-backend/src/pearl/group.rs

```diff
@@ -254,34 +256,42 @@ impl Group {
 
     pub async fn get(&self, key: BobKey) -> Result<BobData, Error> {
         let holders = self.holders.read().await;
-        let mut has_error = false;
         let mut results = vec![];
+        let mut max_delete_ts = 0; // In case of multiple holders with same ts and delete action
+        let mut has_error = false;
         for holder in holders
             .iter_possible_childs_rev(&Key::from(key))
             .map(|(_, x)| &x.data)
         {
-            let get = Self::get_common(&holder, key).await;
-            match get {
-                Ok(data) => {
-                    trace!("get data: {:?} from: {:?}", data, holder);
-                    results.push(data);
-                }
-                Err(err) => {
-                    if err.is_key_not_found() {
-                        debug!("{} not found in {:?}", key, holder)
-                    } else {
+            if holder.end_timestamp() > max_delete_ts {
```

Comment:
This is very interesting and at the same time dangerous optimization. In the current implementation `end_timestamp` calculated as `start_timestamp + self.settings.timestamp_period_as_secs()`. If `timestamp_period` is changed in the config, then the reading logic may start skipping some entries.

I think we can extend this optimization to all records (not just deleted once), but for safety reasons also should increase the interval between record's timestamp and holder's timestamp by `5 * timestamp_period`: 
```
if holder.end_timestamp() + TIMESTAMP_GAP_FOR_HOLDERS_ACCESS * settings.timestamp_period() > max_record_ts { 
```

Proper implementation requires keeping track min and max timestamp for every holder. I've added an issue for the future: https://github.com/qoollo/bob/issues/694

Reply:
I've removed this optimization for now for safety reasons

Reply:
Ok